source/simulator/helper_functions.cc

```diff
@@ -385,9 +385,64 @@ namespace aspect
   Simulator<dim>::
   compute_pressure_scaling_factor()
   {
+    const QMidpoint<dim> quadrature_formula;
+
+    FEValues<dim> fe_values (*mapping,
+                             finite_element,
+                             quadrature_formula,
+                             update_values |
+                             update_gradients |
+                             update_quadrature_points |
+                             update_JxW_values);
+
+    const unsigned int n_q_points = quadrature_formula.size();
+
+    double local_integrated_viscosity = 0.;
+    double local_volume = 0.;
+
+    MaterialModel::MaterialModelInputs<dim> in(n_q_points,
+                                               introspection.n_compositional_fields);
+    MaterialModel::MaterialModelOutputs<dim> out(n_q_points,
+                                                 introspection.n_compositional_fields);
+
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      if (cell->is_locally_owned())
+        {
+          fe_values.reinit (cell);
+          in.reinit(fe_values,
+                    cell,
+                    introspection,
+                    solution);
+
+          // We do not average the output, because  it is not
+          // necessary for a single quadrature point of QMidpoint
+          // (and it fails for Q1 averaging)
+          material_model->evaluate(in, out);
+
+          // Evaluate viscosity at each quadrature point and
+          // calculate the harmonic average
+          for (unsigned int q=0; q<n_q_points; ++q)
+            {
+              Assert(out.viscosities[q] > 0,
+                     ExcMessage ("The viscosity needs to be a "
+                                 "non-negative quantity."));
+
+              local_integrated_viscosity += fe_values.JxW(q)/out.viscosities[q];
+              local_volume += fe_values.JxW(q);
+            }
+        }
+
+    const double integrated_viscosity
+      = Utilities::MPI::sum (local_integrated_viscosity, mpi_communicator);
+
+    const double global_volume
+      = Utilities::MPI::sum (local_volume, mpi_communicator);
```

Comment:
It doesn't really matter, but you can put the two variables together in an array, sum them in one step and with one communication, and then take them apart again. We do this in most other places where we compute an average.